### PR TITLE
Build and Push Sandbox Docker Image

### PR DESCRIPTION
**Pull request by EngHabu**

### CODE REVIEW

File: rsts/tutorials/first_run.rst [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+Getting Started with Flyte
+############################################
+
+.. rubric:: Estimated time to complete: 2 minutes.
+
+Flyte enables scalable, reproducable and reliable orchestration of massively large workflows. In order to get a sense of the product, we have packaged a minimalist version of the Flyte system into a Docker image.
+
+With `docker installed <https://docs.docker.com/get-docker/>`__, run this command: ::
+
+  docker network create flyte-sandbox | docker run --network flyte-sandbox --rm --privileged -p 30081:30081 ghcr.io/flyteorg/flyte-sandbox
```
Comment (EngHabu) [inline rsts/tutorials/first_run.rst:14]: @jeevb what should this command be now?

File: rsts/tutorials/first_run.rst [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+Once the container is ready, it'll output the Console URL. Go ahead and visit that to check out the Flyte UI.
```
Comment (jeevb) [inline rsts/tutorials/first_run.rst:16]:
```suggestion
  docker exec $(docker run -d --rm --name flyte-sandbox --privileged -p 30081:30081 ghcr.io/flyteorg/flyte-sandbox) wait-for-flyte.sh

This creates a container named `flyte-sandbox` within which a sandbox Flyte deployment will be created. Once the deployment is ready, the above command should exit successfully, and the Flyte UI should be available at http://localhost:30081/console.
```

File: docker/sandbox/wait-for-flyte.sh [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+until k3s kubectl explain deployment &> /dev/null; do sleep 1; done
+
+# Wait for Flyte namespace to be created. This is necessary for the next step.
+timeout 600 sh -c "until k3s kubectl get namespace flyte &> /dev/null; do sleep 1; done"  || ( echo >&2 "Timed out while waiting for the Flyte namespace to be created"; exit 1 )
+
+# Wait for flyte deployment
+k3s kubectl wait --for=condition=available deployment/datacatalog deployment/flyteadmin deployment/flyteconsole deployment/flytepropeller -n flyte --timeout=10m || ( echo >&2 "Timed out while waiting for the Flyte deployment to start"; exit 1 )
+
+# Wait for envoy proxy to become ready
+timeout 600 sh -c 'until [[ $(k3s kubectl get daemonset envoy -n flyte -o jsonpath="{.status.numberReady}") -eq 1 ]]; do sleep 1; done' || ( echo >&2 "Timed out while waiting for the Flyte envoy proxy to start"; exit 1 )
```
Comment (jeevb) [inline docker/sandbox/wait-for-flyte.sh:17]:
```suggestion
timeout 600 sh -c 'until [[ $(k3s kubectl get daemonset envoy -n flyte -o jsonpath="{.status.numberReady}") -eq 1 ]]; do sleep 1; done' || ( echo >&2 "Timed out while waiting for the Flyte envoy proxy to start"; exit 1 )

echo "Flyte is ready! Flyte UI is available at http://localhost:30081/console."
```

File: docker/sandbox/flyte-entrypoint-dind.sh [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+k3s server --docker --no-deploy=traefik --no-deploy=servicelb --no-deploy=local-storage --no-deploy=metrics-server &> /var/log/k3s.log &
+K3S_PID=$!
+timeout 600 sh -c "until k3s kubectl explain deployment &> /dev/null; do sleep 1; done" || ( echo >&2 "Timed out while waiting for the Kubernetes cluster to start"; exit 1 )
+echo "Done."
+
+# Deploy flyte
+echo "Deploying Flyte..."
+k3s kubectl apply -f /flyteorg/share/flyte_generated.yaml
+wait-for-flyte.sh
+echo "Done."
```
Comment (jeevb) [inline docker/sandbox/flyte-entrypoint-dind.sh:35]:
```suggestion
```

File: docker/sandbox/flyte-entrypoint-default.sh [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+k3s server --no-deploy=traefik --no-deploy=servicelb --no-deploy=local-storage --no-deploy=metrics-server &> /var/log/k3s.log &
+K3S_PID=$!
+timeout 600 sh -c "until k3s kubectl explain deployment &> /dev/null; do sleep 1; done" || ( echo >&2 "Timed out while waiting for the Kubernetes cluster to start"; exit 1 )
+echo "Done."
+
+# Deploy flyte
+echo "Deploying Flyte..."
+k3s kubectl apply -f /flyteorg/share/flyte_generated.yaml
+wait-for-flyte.sh
+echo "Done."
```
Comment (jeevb) [inline docker/sandbox/flyte-entrypoint-default.sh:18]:
```suggestion
```

File: docker/sandbox/wait-for-flyte.sh [hunk 6]
```diff
@@ -0,0 +1,19 @@
+#!/bin/sh
+
+set -e
+
```
Comment (jeevb) [inline docker/sandbox/wait-for-flyte.sh:4]:
```suggestion

echo "Waiting for Flyte to become ready..."
```

File: docker/sandbox/wait-for-flyte.sh [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+echo "Waiting for Flyte to become ready..."
```
Comment (jeevb) [inline docker/sandbox/wait-for-flyte.sh:5]:
```suggestion
echo "Waiting for Flyte to become ready..."

```

**Comment by v01dXYZ** [comment]
Hi,
Why the Dockerfile is shipping buildkit-for-k8s ?
Is there any reason to prefer it over a `docker build` (which is buildkit too) provided by `dind` ?

**Comment by jeevb** [comment]
> Hi,
> Why does the Dockerfile ship `buildkit-for-k8s` ?
> Is there any reason to prefer it over a `docker build` (which is buildkit too) provided by `dind` ?

You can use the `ghcr.io/flyteorg/flyte-sandbox:dind` image, if you prefer. We're pushing both as options for users.

**Comment by v01dXYZ** [comment]
Thanks for the pointers. I was also trying to understand how `k3s` could run in a container.

I  didn't see the `--privileged` option and it seems it is required by the agent (eg setting the IP tables, snapshotting with overlay2, accessing to `/dev/kmsg`).

But looking at the code of `k3d`, it seems that's also the same thing that's done: https://github.com/rancher/k3d/blob/dbf3ff81721de5837261a0a9b84c6982d1c3966e/pkg/runtimes/docker/translate.go#L90

So, for something a little more thorough than a 5mn  walk-around, it is better to use directly `k3s`.

**Comment by jeevb** [comment]
Right you do have to run the container with `--privileged`. See here: https://docs.flyte.org/en/latest/tutorials/first_run.html. The sandbox should work fine for playing around with Flyte, but not for any production workloads. There are example overlays for running on EKS or GKE (https://github.com/flyteorg/flyte/tree/master/deployment), or you can compose your own. You should also be able to get faster support via Slack: https://docs.google.com/forms/d/e/1FAIpQLScWPJZncyL-bBCWbuQ9HWuMzA8-r2RJc__CSWJoAYUQADN-BQ/viewform.

**Pull request merged**